machine_learning/perceptron.py

```diff
@@ -0,0 +1,112 @@
+import numpy as np
+import pandas as pd
+from sklearn import datasets
+from sklearn.metrics import accuracy_score
+from sklearn.model_selection import train_test_split
+
+
+"""
+Implementation of a basic Perceptron for LINEARLY solvable problems.
+Input data set: The input data set must be 2-dimensional with [0,1] labels.
+Output: The perceptron maps a test set and calculates overall acuraciy.
+"""
+
+
+def usf(val):
+    """Unit Step Function"""
+    return 1 if val >= 0 else 0
+
+
+def sign(val):
+    """Sign Function"""
+    return 1 if val >= 0 else -1
+
+
+class Perceptron:
+    def __init__(
+        self,
+        eta: float = 0.01,
+        epochs: int = 100,
+        bias: int = 1,
+        random_state: int = 42,
+    ) -> None:
+        self.eta = eta
+        self.epochs = epochs
+        self.bias = bias
+        self.errors = []
+        self.theta = None
+        self.activation_function = None
+
+    def __repr__(self) -> str:
+        return (
+            f"Perceptron -> eta: {self.eta} - epocs: {self.epochs} - bias: {self.bias}"
+        )
+
+    def fit(self, X, y):
+        """fitting the perceptron on given data features X (train set) and labels Y to calculate error on"""
+        self.activation_function = usf if min(set(y)) == 0 else sign
+        self.theta = np.zeros(X.shape[1])
+
+        for _ in range(self.epochs):
+            # errors per epoch
+            n_wrong = 0
+
+            for idx, x_i in enumerate(X):
+                # calculate dot product WX
+                y_hat = self.activation_function(np.dot(x_i, self.theta) + self.bias)
+
+                # calculating error
+                error = y[idx] - y_hat
+                if error:
+                    # updating weights and bias
+                    self.theta += self.eta * error * x_i
+                    self.bias += self.eta * error
+                    n_wrong += 1
+
+            self.errors.append(n_wrong)
+
+    def predict(self, X) -> np.array:
```

Comment:
Run the pre-commit for getting to know about the fixes before creating a PR !

Reply:
Do this with pipx instead of pip...
% `pipx install pre-commit`
% `pre-commit install`

pre-commit will install ruff, black, etc. in its own dedicated virtual environment.